libazureinit/src/config.rs

```diff
@@ -0,0 +1,1004 @@
+                base_config_path
+            );
+
+            let additional_toml = Self::load_toml_file(base_config_path)?;
+            base_toml = Self::merge_toml(base_toml, additional_toml);
+            tracing::debug!(
+                "Base TOML after merging azure-init.toml: {:?}",
+                base_toml
+            );
+        }
+
+        let d_dir = dir.join("azure-init.d");
+        if d_dir.is_dir() {
+            let mut toml_files: Vec<_> = fs::read_dir(d_dir)?
+                .filter_map(|entry| {
+                    let entry = entry.ok()?;
+                    let path = entry.path();
+                    if path.extension()?.to_str()? == "toml" {
+                        Some(path)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            toml_files.sort();
+
+            for file_path in toml_files {
+                tracing::info!("Merging configuration from {:?}", file_path);
+                let additional_toml = Self::load_toml_file(file_path)?;
+                base_toml = Self::merge_toml(base_toml, additional_toml); //
+            }
+        }
+
+        Config::from_toml(base_toml)
+    }
+
+    /// Converts a TOML `Value` into a `Config` object.
+    fn from_toml(toml: toml::Value) -> Result<Config, Error> {
+        toml::from_str(&toml.to_string()).map_err(|e| {
+            tracing::error!("Failed to convert TOML to Config: {:?}", e);
+            Error::Io(std::io::Error::new(
+                std::io::ErrorKind::InvalidData,
+                "Failed to convert TOML to Config.",
+            ))
+        })
+    }
+
+    /// Loads a TOML file and returns it as a TOML `Value`.
+    fn load_toml_file(file_path: PathBuf) -> Result<toml::Value, Error> {
+        let content = fs::read_to_string(file_path).map_err(|e| {
+            tracing::error!("Failed to read file: {:?}", e);
+            Error::Io(e)
+        })?;
+
+        toml::from_str(&content).map_err(|e| {
+            tracing::error!("Failed to parse TOML: {:?}", e);
+            Error::Io(std::io::Error::new(
+                std::io::ErrorKind::InvalidData,
+                format!("Failed to parse TOML: {:?}", e),
+            ))
+        })
+    }
+
+    /// Merges two TOML `Value` objects recursively.
+    ///
+    /// This method combines two TOML tables, with values from `additional` taking precedence
+    /// over values from `base`. Nested tables are merged recursively.
+    ///
+    /// # Arguments
+    ///
+    /// * `base` - The base TOML table to merge into
+    /// * `additional` - The additional TOML table whose values take precedence
+    fn merge_toml(
+        mut base: toml::Value,
+        additional: toml::Value,
+    ) -> toml::Value {
+        if let (Some(base_table), Some(additional_table)) =
+            (base.as_table_mut(), additional.as_table())
+        {
+            for (key, additional_value) in additional_table {
```

Comment:
As discussed offline, I think this implementation will have some issues when the users provide bad input. Consider the example:
```
base:
[foo]
a=b
[bar]
c=d

additional:
foo=a

The whole section [foo] of base will be overwritten
```

We can probably address this by ensuring that additional is a valid azure-init's config file before merging. However, my recommendation is to use an existing well known and proven library like [figment](https://crates.io/crates/figment)